### DISCUSSION TITLE
Best way to handle deploying a new worker version and DO NOT kill in-flight tasks?

### DISCUSSION THREAD

**Discussion by ztoolsgithub** [taskforcesh/bullmq#1531] [Q&A]
When deploying new code changes to the workers, we automatically deploy and promote the new version to production on Google App Engine. This is bad since it kills the previous version of the worker in the middle of operations.

1) What's the best way to prevent this behavior and wait until all workers finish their in-flight tasks?
2) I guess an obvious solution would be to not promote the new version until finishing all the in-flight tasks from the current workers, but this has a challenge... How to make sure the workers finish inflight tasks but do not start processing new tasks? I guess pausing the worker could help, but not sure what's the best way to do it.

Also, in case a worker fails to process a task successfully, what's the best practice? Throwing an exception so the task is marked as invalid?

**Comment by manast** [comment]
The normal behavior of ```await worker.close()``` is to wait for the current jobs to finalize before closing. So you must properly close your workers when your Google App engine closes down the app.

**Reply by manast** [reply]
I am no familiar with Google App Engine but probably they are killing your app by sending "SIGTERM" or a similar signal, so you can just listen to that event using ```process.on("SIGTERM", ...)``` and close your workers there.

**Reply by ztoolsgithub** [reply]
hi @manast! Looks like they automatically kill the service when there is a new release or the service becomes idle (on autoscaling mode): https://cloud.google.com/appengine/docs/standard/how-instances-are-managed

So it doesn't seem to me that I'd be having a way to close the worker if that's the case. What could it be the best way to handle it then? Prob not auto-promoting the new version will avoid killing the job, but we'd need a way to safely stop the workers 🤔 Any ideas?

![Screen Shot 2022-11-11 at 1 18 31 AM](https://user-images.githubusercontent.com/116674820/201233071-8a7e25d6-7a5e-44ab-a975-a9e1971259f4.png)

**Reply by manast** [reply]
I think you should have 30 seconds to gracefully close your workers according to the documentation you posted. So if the jobs are fast enough it should work by adding a listener to SIGTERM as I mentioned above. 
https://cloud.google.com/appengine/docs/flexible/custom-runtimes/build#application_shutdown
